### PR TITLE
Fix the memory usage issue of logits in generate()

### PR DESCRIPTION
**Pull request by kjohew**
# What does this PR do?

As discussed in [#34796](https://github.com/huggingface/transformers/issues/34796), logits should be sliced before being copied to avoid unnecessary memory usage, especially during the first call to generate.

## Before submitting
- [x] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?
@zucchini-nlp

### CODE REVIEW

**Review by zucchini-nlp — APPROVED** [review]
Thanks! Failing tests are not related, re-triggered them in case it was flaky

**Review by ArthurZucker — APPROVED** [review]
Thanks 🤗
